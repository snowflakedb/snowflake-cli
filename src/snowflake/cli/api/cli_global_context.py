# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field, replace
from functools import wraps
from pathlib import Path
from typing import TYPE_CHECKING, Iterator, Optional

from snowflake.cli.api.connections import ConnectionContext, OpenConnectionCache
from snowflake.cli.api.exceptions import MissingConfiguration
from snowflake.cli.api.metrics import CLIMetrics
from snowflake.cli.api.output.formats import OutputFormat
from snowflake.cli.api.rendering.jinja import CONTEXT_KEY
from snowflake.connector import SnowflakeConnection

if TYPE_CHECKING:
    from snowflake.cli.api.project.definition_manager import DefinitionManager
    from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition

_CONNECTION_CACHE = OpenConnectionCache()


@dataclass
class _CliGlobalContextManager:
    connection_context: ConnectionContext = field(default_factory=ConnectionContext)
    connection_cache: OpenConnectionCache = (
        _CONNECTION_CACHE  # by default, use global cache
    )

    output_format: OutputFormat = OutputFormat.TABLE
    silent: bool = False
    verbose: bool = False
    experimental: bool = False
    enable_tracebacks: bool = True

    metrics: CLIMetrics = field(default_factory=CLIMetrics)

    project_path_arg: str | None = None
    project_is_optional: bool = True
    project_env_overrides_args: dict[str, str] = field(default_factory=dict)

    # FIXME: this property only exists to help implement
    # nativeapp_definition_v2_to_v1 and single_app_and_package.
    # Consider changing the way this calculation is provided to commands
    # in order to remove this logic (then make project_definition a non-cloned @property)
    override_project_definition: ProjectDefinition | None = None

    _definition_manager: DefinitionManager | None = None

    # which properties invalidate our current DefinitionManager?
    DEFINITION_MANAGER_DEPENDENCIES = [
        "project_path_arg",
        "project_is_optional",
        "project_env_overrides_args",
    ]

    def reset(self):
        self.__init__()

    def clone(self) -> _CliGlobalContextManager:
        return replace(
            self,
            connection_context=self.connection_context.clone(),
            project_env_overrides_args=self.project_env_overrides_args.copy(),
            metrics=self.metrics.clone(),
        )

    def __setattr__(self, prop, val):
        if prop in self.DEFINITION_MANAGER_DEPENDENCIES:
            self._clear_definition_manager()

        super().__setattr__(prop, val)

    @property
    def project_definition(self) -> ProjectDefinition | None:
        if self.override_project_definition:
            return self.override_project_definition

        return self._definition_manager_or_raise().project_definition

    @property
    def project_root(self) -> Path:
        return Path(self._definition_manager_or_raise().project_root)

    @property
    def template_context(self) -> dict:
        return self._definition_manager_or_raise().template_context

    @property
    def connection(self) -> SnowflakeConnection:
        """
        Returns a connection for our configured context from the configured cache.
        By default, this is the global _CONNECTION_CACHE. If a matching connection
        does not already exist, creates a new connection and caches it.
        """
        self.connection_context.validate_and_complete()
        return self.connection_cache[self.connection_context]

    def _definition_manager_or_raise(self) -> DefinitionManager:
        """
        (Re-)parses project definition based on project args (project_path_arg and
        project_env_overrides_args). If we cannot provide a project definition
        (i.e. no snowflake.yml) and require one, raises MissingConfiguration.
        """
        from snowflake.cli.api.project.definition_manager import DefinitionManager

        # don't need to re-parse definition if we already have one
        if not self._definition_manager:
            dm = DefinitionManager(
                self.project_path_arg,
                {CONTEXT_KEY: {"env": self.project_env_overrides_args}},
            )
            if not dm.has_definition_file and not self.project_is_optional:
                raise MissingConfiguration(
                    "Cannot find project definition (snowflake.yml). Please provide a path to the project or run this command in a valid project directory."
                )
            self._definition_manager = dm

        return self._definition_manager

    def _clear_definition_manager(self):
        """
        Force re-calculation of definition_manager and its dependent attributes
        (template_context, project_definition, and project_root).
        """
        self._definition_manager = None


class _CliGlobalContextAccess:
    def __init__(self, manager: _CliGlobalContextManager):
        self._manager = manager

    @property
    def connection(self) -> SnowflakeConnection:
        return self._manager.connection

    @property
    def connection_context(self) -> ConnectionContext:
        return self._manager.connection_context

    @property
    def enable_tracebacks(self) -> bool:
        return self._manager.enable_tracebacks

    @property
    def metrics(self) -> CLIMetrics:
        return self._manager.metrics

    @property
    def output_format(self) -> OutputFormat:
        return self._manager.output_format

    @property
    def verbose(self) -> bool:
        return self._manager.verbose

    @property
    def experimental(self) -> bool:
        return self._manager.experimental

    @property
    def project_definition(self) -> ProjectDefinition | None:
        return self._manager.project_definition

    @property
    def project_root(self) -> Path | None:
        return self._manager.project_root

    @property
    def template_context(self) -> dict:
        return self._manager.template_context

    @property
    def silent(self) -> bool:
        if self._should_force_mute_intermediate_output:
            return True
        return self._manager.silent

    @property
    def _should_force_mute_intermediate_output(self) -> bool:
        """Computes whether cli_console output should be muted."""
        return self._manager.output_format == OutputFormat.JSON

    @property
    def snow_api_root(
        self,
    ) -> Optional[
        object
    ]:  # Should be Optional[Root], but we need local import for performance reasons
        from snowflake.core import Root

        if self.connection:
            return Root(self.connection)
        else:
            return None


_CLI_CONTEXT_MANAGER: ContextVar[_CliGlobalContextManager | None] = ContextVar(
    "cli_context", default=None
)


def get_cli_context_manager() -> _CliGlobalContextManager:
    mgr = _CLI_CONTEXT_MANAGER.get()
    if not mgr:
        mgr = _CliGlobalContextManager()
        _CLI_CONTEXT_MANAGER.set(mgr)
    return mgr


def get_cli_context() -> _CliGlobalContextAccess:
    return _CliGlobalContextAccess(get_cli_context_manager())


def span(span_name: str):
    """
    Decorator to start a command metrics span that encompasses a whole function

    Must be used instead of directly calling @get_cli_context().metrics.span(span_name)
    as a decorator to ensure that the cli context is grabbed at run time instead of at
    module load time, which would not reflect forking
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with get_cli_context().metrics.span(span_name):
                return func(*args, **kwargs)

        return wrapper

    return decorator


@contextmanager
def fork_cli_context(
    connection_overrides: dict | None = None,
    project_env_overrides: dict[str, str] | None = None,
    project_is_optional: bool | None = None,
    project_path: str | None = None,
) -> Iterator[_CliGlobalContextAccess]:
    """
    Forks the global CLI context, making changes that are only visible
    (e.g. via get_cli_context()) while inside this context manager.

    Please note that environment variable changes are only visible through
    the project definition; os.getenv / os.environ / get_env_value are not
    affected by these new values.
    """
    old_manager = get_cli_context_manager()
    new_manager = old_manager.clone()
    token = _CLI_CONTEXT_MANAGER.set(new_manager)

    if connection_overrides:
        new_manager.connection_context.update(**connection_overrides)

    if project_env_overrides:
        new_manager.project_env_overrides_args.update(project_env_overrides)

    if project_is_optional is not None:
        new_manager.project_is_optional = project_is_optional

    if project_path:
        new_manager.project_path_arg = project_path

    yield _CliGlobalContextAccess(new_manager)
    _CLI_CONTEXT_MANAGER.reset(token)
