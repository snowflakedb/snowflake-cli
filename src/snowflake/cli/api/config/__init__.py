# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Configuration package with support for legacy and next-generation config systems.

This package provides a feature flag mechanism to switch between the legacy
configuration system and the new config-ng system using the environment variable
SNOWFLAKE_CLI_CONFIG_NG.

Usage:
    # Enable new config system
    export SNOWFLAKE_CLI_CONFIG_NG=1

    # Use legacy config system (default)
    unset SNOWFLAKE_CLI_CONFIG_NG

Environment Variables:
    SNOWFLAKE_CLI_CONFIG_NG: Set to '1', 'true', 'yes', or 'on' to enable config-ng
"""

from __future__ import annotations

import os
from typing import Literal

# Environment variable to enable the new config-ng system
CONFIG_NG_ENV_VAR: Literal["SNOWFLAKE_CLI_CONFIG_NG"] = "SNOWFLAKE_CLI_CONFIG_NG"


def _is_config_ng_enabled() -> bool:
    """
    Check if the new config-ng system should be used.

    Returns:
        True if config-ng should be used, False for legacy system
    """
    return os.environ.get(CONFIG_NG_ENV_VAR, "").lower() in ("1", "true", "yes", "on")


# Conditionally import from legacy or config-ng based on environment variable
if _is_config_ng_enabled():
    # Import ONLY from the new config-ng system
    # This will fail if config-ng is not fully implemented, which is intentional
    # to help identify what needs to be implemented
    try:
        from .config_ng import *  # noqa: F403, F401
    except ImportError as e:
        # If config-ng is not fully implemented, fail fast with a clear error
        raise ImportError(
            f"Config-ng is enabled via {CONFIG_NG_ENV_VAR} but is not fully implemented. "
            f"Missing implementation: {e}. "
            f"Either implement the missing functionality in config_ng.py or "
            f"disable config-ng by unsetting {CONFIG_NG_ENV_VAR}."
        ) from e
else:
    # Use the legacy configuration system
    from .legacy import *  # noqa: F403, F401

# Export the feature flag utilities and factory method for external use
__all__ = ["_is_config_ng_enabled", "CONFIG_NG_ENV_VAR", "get_config_manager"]
