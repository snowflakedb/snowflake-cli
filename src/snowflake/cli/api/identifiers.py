# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import re
from pathlib import Path

from click import ClickException
from snowflake.cli.api.exceptions import FQNInconsistencyError, FQNNameError
from snowflake.cli.api.project.schemas.v1.identifier_model import (
    ObjectIdentifierBaseModel,
)
from snowflake.cli.api.project.util import VALID_IDENTIFIER_REGEX, identifier_for_url


class FQN:
    """
    Class representing an identifier and supporting fully qualified names.

    The instance supports builder pattern that allows updating the identifier with database and
    schema from different sources. For example:

    fqn = FQN.from_string("my_schema.object").using_connection(conn)
    fqn = FQN.from_identifier_model(cli_context.project_definition.streamlit).using_context()
    fqn = FQN.from_string("my_name").set_database("db").set_schema("foo")
    """

    def __init__(
        self,
        database: str | None,
        schema: str | None,
        name: str,
        signature: str | None = None,
    ):
        self._database = database
        self._schema = schema
        self._name = name
        self.signature = signature

    @property
    def database(self) -> str | None:
        return self._database

    @property
    def schema(self) -> str | None:
        return self._schema

    @property
    def name(self) -> str:
        return self._name

    @property
    def prefix(self) -> str:
        if self.database:
            return f"{self.database}.{self.schema if self.schema else 'PUBLIC'}"
        if self.schema:
            return f"{self.schema}"
        return ""

    @property
    def identifier(self) -> str:
        if self.prefix:
            return f"{self.prefix}.{self.name}"
        return self.name

    @property
    def url_identifier(self) -> str:
        return ".".join(identifier_for_url(part) for part in self.identifier.split("."))

    @property
    def sql_identifier(self) -> str:
        if self.signature:
            return f"IDENTIFIER('{self.identifier}'){self.signature}"
        return f"IDENTIFIER('{self.identifier}')"

    def __str__(self):
        return self.identifier

    def __eq__(self, other):
        return self.identifier == other.identifier

    @classmethod
    def from_string(cls, identifier: str) -> "FQN":
        """
        Takes in an object name in the form [[database.]schema.]name. Returns a FQN instance.
        """
        qualifier_pattern = rf"(?:(?P<first_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?:(?P<second_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?P<name>{VALID_IDENTIFIER_REGEX})(?P<signature>\(.*\))?"
        result = re.fullmatch(qualifier_pattern, identifier)

        if result is None:
            raise FQNNameError(identifier)

        unqualified_name = result.group("name")
        if result.group("second_qualifier") is not None:
            database = result.group("first_qualifier")
            schema = result.group("second_qualifier")
        else:
            database = None
            schema = result.group("first_qualifier")

        signature = None
        if result.group("signature"):
            signature = result.group("signature")
        return cls(
            name=unqualified_name, schema=schema, database=database, signature=signature
        )

    @classmethod
    def from_stage(cls, stage: str) -> "FQN":
        name = stage
        if stage.startswith("@"):
            name = stage[1:]
        if stage.startswith("snow://"):
            name = stage[7:]
        if stage.startswith("~"):
            return cls(name="~", database=None, schema=None)
        return cls.from_string(name)

    @classmethod
    def from_stage_path(cls, stage_path: str) -> "FQN":
        stage = Path(stage_path).parts[0]
        return cls.from_stage(stage)

    @classmethod
    def from_identifier_model_v1(cls, model: ObjectIdentifierBaseModel) -> "FQN":
        """Create an instance from object model."""
        if not isinstance(model, ObjectIdentifierBaseModel):
            raise ClickException(
                f"Expected {type(ObjectIdentifierBaseModel).__name__}, got {model}."
            )

        fqn = cls.from_string(model.name)

        if fqn.database and model.database:
            raise FQNInconsistencyError("database", model.name)
        if fqn.schema and model.schema_name:
            raise FQNInconsistencyError("schema", model.name)

        return fqn.set_database(model.database).set_schema(model.schema_name)

    @classmethod
    def from_identifier_model_v2(cls, model) -> "FQN":
        """Create an instance from object model."""
        from snowflake.cli.api.project.schemas.entities.common import Identifier

        if not isinstance(model, Identifier):
            raise ClickException(f"Expected {type(Identifier).__name__}, got {model}.")

        fqn = cls.from_string(model.name)

        if fqn.database and model.database:
            raise FQNInconsistencyError("database", model.name)
        if fqn.schema and model.schema_:
            raise FQNInconsistencyError("schema", model.name)

        return fqn.set_database(model.database).set_schema(model.schema_)

    def set_database(self, database: str | None) -> "FQN":
        if database:
            self._database = database
        return self

    def set_schema(self, schema: str | None) -> "FQN":
        if schema:
            self._schema = schema
        return self

    def set_name(self, name: str) -> "FQN":
        self._name = name
        return self

    def using_connection(self, conn) -> "FQN":
        """Update the instance with database and schema from connection."""
        # Update the identifier only it if wasn't already a qualified name
        if conn.database and not self.database:
            self.set_database(conn.database)
        if conn.schema and not self.schema:
            self.set_schema(conn.schema)
        return self

    def using_context(self) -> "FQN":
        """Update the instance with database and schema from connection in current cli context."""
        from snowflake.cli.api.cli_global_context import get_cli_context

        return self.using_connection(get_cli_context().connection)

    def to_dict(self) -> dict:
        return {"name": self.name, "schema": self.schema, "database": self.database}
