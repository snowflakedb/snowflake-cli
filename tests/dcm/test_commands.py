from unittest import mock

import pytest
from snowflake.cli.api.identifiers import FQN

DCMProjectManager = "snowflake.cli._plugins.dcm.commands.DCMProjectManager"
ObjectManager = "snowflake.cli._plugins.dcm.commands.ObjectManager"
get_entity_for_operation = (
    "snowflake.cli._plugins.dcm.commands.get_entity_for_operation"
)


@pytest.fixture
def mock_project_exists():
    with mock.patch(
        "snowflake.cli._plugins.dcm.commands.ObjectManager.object_exists",
        return_value=True,
    ) as _fixture:
        yield _fixture


@mock.patch(DCMProjectManager)
@mock.patch(ObjectManager)
def test_create(mock_om, mock_pm, runner, project_directory):
    mock_om().object_exists.return_value = False
    with project_directory("dcm_project"):
        command = ["dcm", "create"]
        result = runner.invoke(command)
        assert result.exit_code == 0, result.output

        mock_pm().create.assert_called_once()
        create_kwargs = mock_pm().create.mock_calls[0].kwargs
        assert create_kwargs["project"].fqn == FQN.from_string("my_project")


@mock.patch(DCMProjectManager)
@mock.patch(ObjectManager)
@pytest.mark.parametrize("if_not_exists", [False, True])
def test_create_object_exists(
    mock_om, mock_pm, runner, project_directory, if_not_exists
):
    mock_om().object_exists.return_value = True
    with project_directory("dcm_project"):
        command = ["dcm", "create"]
        if if_not_exists:
            command.append("--if-not-exists")
        result = runner.invoke(command)
        if if_not_exists:
            assert result.exit_code == 0, result.output
            assert "DCM Project 'my_project' already exists." in result.output
        else:
            assert result.exit_code == 1, result.output

        mock_pm().create.assert_not_called()


@mock.patch(DCMProjectManager)
def test_deploy_project(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(["dcm", "deploy", "fooBar", "--from", "@my_stage"])
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration=None,
        from_stage="@my_stage",
        variables=None,
        alias=None,
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_deploy_project_with_from_stage(
    mock_pm, runner, project_directory, mock_cursor
):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(["dcm", "deploy", "fooBar", "--from", "@my_stage"])
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration=None,
        from_stage="@my_stage",
        variables=None,
        alias=None,
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_deploy_project_with_variables(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        ["dcm", "deploy", "fooBar", "--from", "@my_stage", "-D", "key=value"]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration=None,
        from_stage="@my_stage",
        variables=["key=value"],
        alias=None,
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_deploy_project_with_configuration(
    mock_pm, runner, project_directory, mock_cursor
):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        [
            "dcm",
            "deploy",
            "fooBar",
            "--from",
            "@my_stage",
            "--configuration",
            "some_configuration",
        ]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration="some_configuration",
        from_stage="@my_stage",
        variables=None,
        alias=None,
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_deploy_project_with_alias(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        ["dcm", "deploy", "fooBar", "--from", "@my_stage", "--alias", "my_alias"]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration=None,
        from_stage="@my_stage",
        variables=None,
        alias="my_alias",
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_plan_project(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        [
            "dcm",
            "plan",
            "fooBar",
            "--from",
            "@my_stage",
            "-D",
            "key=value",
            "--configuration",
            "some_configuration",
        ]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration="some_configuration",
        from_stage="@my_stage",
        dry_run=True,
        variables=["key=value"],
        output_path=None,
    )


@mock.patch(DCMProjectManager)
def test_plan_project_with_from_stage(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        [
            "dcm",
            "plan",
            "fooBar",
            "--from",
            "@my_stage",
            "-D",
            "key=value",
            "--configuration",
            "some_configuration",
        ]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration="some_configuration",
        from_stage="@my_stage",
        dry_run=True,
        variables=["key=value"],
        output_path=None,
    )


def test_list_command_alias(mock_connect, runner):
    result = runner.invoke(
        [
            "object",
            "list",
            "dcm",
            "--like",
            "%PROJECT_NAME%",
            "--in",
            "database",
            "my_db",
        ]
    )

    assert result.exit_code == 0, result.output
    result = runner.invoke(
        ["dcm", "list", "--like", "%PROJECT_NAME%", "--in", "database", "my_db"],
        catch_exceptions=False,
    )
    assert result.exit_code == 0, result.output

    queries = mock_connect.mocked_ctx.get_queries()
    assert len(queries) == 2
    assert (
        queries[0]
        == queries[1]
        == "show DCM Projects like '%PROJECT_NAME%' in database my_db"
    )


@mock.patch(DCMProjectManager)
def test_list_deployments(mock_pm, runner):
    result = runner.invoke(["dcm", "list-deployments", "fooBar"])

    assert result.exit_code == 0, result.output

    mock_pm().list_versions.assert_called_once_with(
        project_name=FQN.from_string("fooBar")
    )


@mock.patch(DCMProjectManager)
@pytest.mark.parametrize("if_exists", [True, False])
def test_drop_version(mock_pm, runner, if_exists):
    command = ["dcm", "drop-deployment", "fooBar", "v1"]
    if if_exists:
        command.append("--if-exists")

    result = runner.invoke(command)

    assert result.exit_code == 0, result.output
    assert "Version 'v1' dropped from DCM Project 'fooBar'" in result.output

    mock_pm().drop_deployment.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        version_name="v1",
        if_exists=if_exists,
    )


@mock.patch(DCMProjectManager)
@pytest.mark.parametrize(
    "version_name,should_warn",
    [
        ("version", True),
        ("VERSION", True),
        ("Version", True),
        ("VERSION$1", False),
        ("v1", False),
        ("my_version", False),
        ("version1", False),
        ("actual_version", False),
    ],
)
def test_drop_version_shell_expansion_warning(
    mock_pm, runner, version_name, should_warn
):
    """Test that warning is displayed for version names that look like shell expansion results."""
    result = runner.invoke(["dcm", "drop-deployment", "fooBar", version_name])

    assert result.exit_code == 0, result.output

    if should_warn:
        assert "might be truncated due to shell expansion" in result.output
        assert "try using single quotes" in result.output
    else:
        assert "might be truncated due to shell expansion" not in result.output

    mock_pm().drop_deployment.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        version_name=version_name,
        if_exists=False,
    )


def test_drop_project(mock_connect, runner):
    result = runner.invoke(
        [
            "object",
            "drop",
            "dcm",
            "my_project",
        ]
    )

    assert result.exit_code == 0, result.output

    result = runner.invoke(
        ["dcm", "drop", "my_project"],
        catch_exceptions=False,
    )
    assert result.exit_code == 0, result.output

    queries = mock_connect.mocked_ctx.get_queries()
    assert len(queries) == 2
    assert queries[0] == queries[1] == "drop DCM Project IDENTIFIER('my_project')"


def test_describe_command_alias(mock_connect, runner):
    result = runner.invoke(
        [
            "object",
            "describe",
            "dcm",
            "PROJECT_NAME",
        ]
    )

    assert result.exit_code == 0, result.output
    result = runner.invoke(
        ["dcm", "describe", "PROJECT_NAME"],
        catch_exceptions=False,
    )
    assert result.exit_code == 0, result.output

    queries = mock_connect.mocked_ctx.get_queries()
    assert len(queries) == 2
    assert queries[0] == queries[1] == "describe DCM Project IDENTIFIER('PROJECT_NAME')"


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_deploy_project_with_sync(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that files are synced to project stage when from_stage is not provided and project definition exists."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "deploy", "my_project"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called
        mock_sync.assert_called_once()

        # Verify that execute was called with the project's stage as from_stage
        call_args = mock_pm().execute.call_args
        assert call_args is not None
        # Since files were synced to the project's stage, that stage should be used as from_stage
        assert call_args.kwargs["project_name"].identifier == "my_project"
        assert call_args.kwargs["from_stage"] == "my_project_stage"


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_plan_project_with_sync(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that files are synced to project stage when from_stage is not provided and project definition exists."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "plan", "my_project"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called
        mock_sync.assert_called_once()

        # Verify that execute was called with the project's stage as from_stage
        call_args = mock_pm().execute.call_args
        assert call_args is not None
        # Since files were synced to the project's stage, that stage should be used as from_stage
        assert call_args.kwargs["project_name"].identifier == "my_project"
        assert call_args.kwargs["from_stage"] == "my_project_stage"


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_deploy_project_with_prune(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that prune flag is passed to sync_artifacts_with_stage when --prune is used."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "deploy", "my_project", "--prune"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called with prune=True
        mock_sync.assert_called_once()
        call_args = mock_sync.call_args
        assert call_args.kwargs["prune"] is True


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_plan_project_with_prune(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that prune flag is passed to sync_artifacts_with_stage when --prune is used."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "plan", "my_project", "--prune"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called with prune=True
        mock_sync.assert_called_once()
        call_args = mock_sync.call_args
        assert call_args.kwargs["prune"] is True


def test_deploy_prune_and_from_mutually_exclusive(runner):
    """Test that --prune and --from flags are mutually exclusive."""
    result = runner.invoke(
        ["dcm", "deploy", "my_project", "--prune", "--from", "@my_stage"]
    )
    assert result.exit_code != 0
    assert "are incompatible and cannot be used" in result.output


def test_plan_prune_and_from_mutually_exclusive(runner):
    """Test that --prune and --from flags are mutually exclusive."""
    result = runner.invoke(
        ["dcm", "plan", "my_project", "--prune", "--from", "@my_stage"]
    )
    assert result.exit_code != 0
    assert "are incompatible and cannot be used" in result.output


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_deploy_project_without_prune(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that prune defaults to False when --prune is not used."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "deploy", "my_project"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called with prune=False (default)
        mock_sync.assert_called_once()
        call_args = mock_sync.call_args
        assert call_args.kwargs["prune"] is False


@mock.patch("snowflake.cli._plugins.dcm.commands.sync_artifacts_with_stage")
@mock.patch(DCMProjectManager)
def test_plan_project_without_prune(
    mock_pm, mock_sync, runner, project_directory, mock_cursor
):
    """Test that prune defaults to False when --prune is not used."""
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    with project_directory("dcm_project"):
        result = runner.invoke(["dcm", "plan", "my_project"])
        assert result.exit_code == 0, result.output

        # Verify that sync was called with prune=False (default)
        mock_sync.assert_called_once()
        call_args = mock_sync.call_args
        assert call_args.kwargs["prune"] is False


@mock.patch(DCMProjectManager)
def test_plan_project_with_output_path(mock_pm, runner, project_directory, mock_cursor):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        [
            "dcm",
            "plan",
            "fooBar",
            "--from",
            "@my_stage",
            "--output-path",
            "@output_stage/results",
        ]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration=None,
        from_stage="@my_stage",
        dry_run=True,
        variables=None,
        output_path="@output_stage/results",
    )


@mock.patch(DCMProjectManager)
def test_plan_project_with_output_path_and_configuration(
    mock_pm, runner, project_directory, mock_cursor
):
    mock_pm().execute.return_value = mock_cursor(rows=[("[]",)], columns=("operations"))

    result = runner.invoke(
        [
            "dcm",
            "plan",
            "fooBar",
            "--from",
            "@my_stage",
            "--configuration",
            "some_config",
            "--output-path",
            "@output_stage",
        ]
    )
    assert result.exit_code == 0, result.output

    mock_pm().execute.assert_called_once_with(
        project_name=FQN.from_string("fooBar"),
        configuration="some_config",
        from_stage="@my_stage",
        dry_run=True,
        variables=None,
        output_path="@output_stage",
    )
