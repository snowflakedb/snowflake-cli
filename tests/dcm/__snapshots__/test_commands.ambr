# serializer version: 1
# name: TestDCMRefresh.test_refresh_with_fresh_tables
  '''
  
  +------------------------------------------------------------------------------+
  | refreshed_tables                                                             |
  |------------------------------------------------------------------------------|
  | [{'dt_name': 'JW_DCM_TESTALL.ANALYTICS.DYNAMIC_EMPLOYEES',                   |
  | 'refreshed_dt_count': 0, 'data_timestamp': '1760356974.543', 'statistics':   |
  | 'No new data'}]                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: TestDCMRefresh.test_refresh_with_no_dynamic_tables
  '''
  
  +------------------+
  | refreshed_tables |
  |------------------|
  | []               |
  +------------------+
  
  '''
# ---
# name: TestDCMRefresh.test_refresh_with_outdated_tables
  '''
  
  +------------------------------------------------------------------------------+
  | refreshed_tables                                                             |
  |------------------------------------------------------------------------------|
  | [{'dt_name': 'JW_DCM_TESTALL.ANALYTICS.DYNAMIC_EMPLOYEES',                   |
  | 'refreshed_dt_count': 1, 'data_timestamp': '1760357032.175', 'statistics':   |
  | '{"insertedRows":5,"copiedRows":0,"deletedRows":5}'}]                        |
  +------------------------------------------------------------------------------+
  
  '''
# ---
